pkg/controller/generic/transform/controller.go

```diff
@@ -241,6 +245,14 @@ func (ctrl *Controller[Input, Output]) processInputs(
 
 		mappedOut := ctrl.mapFunc(in)
 
+		if mappedOut == zeroOutput {
```

Comment:
I think we should go with something like `func isZero[T any](val T) bool { reflect.ValueOf(v).IsZero() }`. I know that using reflect here is not ideal, but it will bring almost no performance penalty for small objects (smaller than 1000 bytes) and we don't need to add new `ComparableResourceWithRD` constraint. On my test IsZero for small objects is around 4-8ns.

Reply:
Actually, thinking this further - maybe we should just use `MapMetadataFunc func(Input) (Output, bool)` or `func(Input) Optional[Output]`?

Reply:
zero here is just `nil`, I don't see why?

Reply:
Yeah, but technically `Controller` make no assumptions about that, that why you can't use `nil` directly. At the same time using `comparable` here looks like a bit of hack, since you are comparing stuff with zero, and not with each other.

Reply:
not sure I get it. all resources from practical point of view are pointers to some type, so zero value is nil, but as I can't use `nil` directly, I use zero value which we use almost everywhere else. so it compiles to `if a == nil`, and that is what we want, rather than going via reflection for real

Reply:
we could do `Optional` as well, but that would break the API

Reply:
reworked to be optional